File: windows/DoNotType.Core.Tests/ProviderRecommendationTests.cs
using DoNotType.Core;
using Xunit;

namespace DoNotType.Core.Tests;

/// <summary>
/// What the settings picker advises, and the invariants that keep the advice honest.
/// </summary>
/// <remarks>
/// The same recommendation is hand-written four times — Swift for macOS and iOS, this file's
/// subject for Windows, Kotlin for Android — and the failure mode of that arrangement is one
/// client quietly advising something the others do not. Matching tests exist in the other two
/// languages, for the same reason `FailureAdviceTests` does.
/// </remarks>
public sealed class ProviderRecommendationTests
{
    /// <summary>
    /// Two, and these two. The list is the product decision; everything below is a consequence.
    /// </summary>
    [Fact]
    public void TheRecommendedSetIsTheTwoEndsOfOneAxis()
    {
        Assert.Equal(new[] { ProviderKind.Gemini, ProviderKind.XAI }, ProviderFactory.Recommended);
        // The axis itself: one reads the screen, the other cannot. Two recommendations that
        // differed on nothing would be a coin toss dressed as advice.
        Assert.False(ProviderKind.Gemini.IsSpeechRecognition());
        Assert.True(ProviderKind.XAI.IsSpeechRecognition());
    }

    /// <summary>A picker that recommends everything recommends nothing.</summary>
    [Fact]
    public void OnlyTheRecommendedTwoCarryANote()
    {
        foreach (var kind in Enum.GetValues<ProviderKind>())
        {
            Assert.Equal(kind.IsRecommended(), kind.RecommendationNote().Length > 0);
        }
    }

    /// <summary>
    /// Order is the recommendation that survives a dropdown too short to show every row, and the
    /// form indexes into this list rather than casting, so a gap or a duplicate here would put the
    /// wrong backend in the settings file.
    /// </summary>
    [Fact]
    public void EveryBackendIsOfferedOnceAndTheRecommendedOnesComeFirst()
    {
        Assert.Equal(Enum.GetValues<ProviderKind>().Length, ProviderFactory.PickerOrder.Count);
        Assert.Equal(
            Enum.GetValues<ProviderKind>().OrderBy(k => k),
            ProviderFactory.PickerOrder.OrderBy(k => k));
        Assert.Equal(ProviderFactory.Recommended, ProviderFactory.PickerOrder.Take(2));
    }

    /// <summary>
    /// The default a fresh install gets has to be one we tell people to pick, or the settings
    /// window is arguing with the installer.
    /// </summary>
    [Fact]
    public void TheDefaultForNewInstallsIsRecommended()
    {
        Assert.True(ProviderFactory.DefaultForNewInstalls.IsRecommended());
    }

    /// <summary>
    /// Advice belongs on the row of a picker. Names that reach history rows, log lines and
    /// connection errors say what ran.
    /// </summary>
    [Fact]
    public void AdviceStaysOutOfTheNameUsedForRecords()
    {
        foreach (var kind in Enum.GetValues<ProviderKind>())
        {
            Assert.DoesNotContain("recommended", kind.DisplayName());
            Assert.StartsWith(kind.DisplayName(), kind.PickerLabel());
        }

        Assert.Equal("Gemini — recommended", ProviderKind.Gemini.PickerLabel());
        Assert.Equal("OpenRouter (gateway — prefer Gemini for Gemini models)", ProviderKind.OpenRouter.PickerLabel());
    }

    /// <summary>
    /// The claim each one is recommended for, in the words the other clients use. A number that
    /// moves in docs/EVALUATION.md has to move in four places, and this is the one that says so.
    /// </summary>
    [Fact]
    public void TheNotesMakeTheMeasuredClaimAndNotAVaguerOne()
    {
        Assert.Contains("reads the screen", ProviderKind.Gemini.RecommendationNote());
        Assert.Contains("44 of 48", ProviderKind.Gemini.RecommendationNote());
        Assert.Contains("cannot see the screen", ProviderKind.XAI.RecommendationNote());
        Assert.Contains("15 of 48", ProviderKind.XAI.RecommendationNote());
    }
}

File: Tests/DoNotTypeCoreTests/ProviderRecommendationTests.swift
import XCTest

@testable import DoNotTypeCore

/// What the settings pickers advise, and the invariants that keep the advice honest.
///
/// These are cheap assertions about strings, which is normally a smell. They are here because the
/// same recommendation is hand-written four times — Swift for macOS and iOS, C# for Windows,
/// Kotlin for Android — and the failure mode of that arrangement is one client quietly advising
/// something the others do not. Matching tests exist in `FailureAdviceTests` for the same reason.
final class ProviderRecommendationTests: XCTestCase {

    /// Two, and these two. The list is the product decision; everything below is a consequence of
    /// it, so a change here should be deliberate enough to update this line.
    func testTheRecommendedSetIsTheTwoEndsOfOneAxis() {
        XCTAssertEqual(ProviderKind.recommended, [.google, .xai])
        // The axis itself: one reads the screen, the other cannot. A recommendation of two
        // backends that differed on nothing would be a coin toss dressed as advice.
        XCTAssertFalse(ProviderKind.google.isSpeechRecognition)
        XCTAssertTrue(ProviderKind.xai.isSpeechRecognition)
    }

    /// A picker that recommends everything recommends nothing.
    func testOnlyTheRecommendedTwoCarryANote() {
        for kind in ProviderKind.allCases {
            XCTAssertEqual(
                kind.recommendationNote != nil, kind.isRecommended,
                "\(kind.rawValue) disagrees about whether it is recommended")
            XCTAssertEqual(kind.ungroundedRecommendationNote != nil, kind.isRecommended)
        }
    }

    /// Order is the recommendation that survives a fixed-height dropdown, so it is asserted rather
    /// than left to whatever order `allCases` happens to have.
    func testEveryBackendIsOfferedAndTheRecommendedOnesComeFirst() {
        XCTAssertEqual(Set(ProviderKind.pickerOrder), Set(ProviderKind.allCases))
        XCTAssertEqual(ProviderKind.pickerOrder.count, ProviderKind.allCases.count)
        XCTAssertEqual(Array(ProviderKind.pickerOrder.prefix(2)), ProviderKind.recommended)
    }

    /// The default a fresh install gets has to be one of the two we tell people to pick, or the
    /// settings window is arguing with the installer.
    func testTheDefaultForNewInstallsIsRecommended() {
        XCTAssertTrue(ProviderKind.defaultForNewInstalls.isRecommended)
        XCTAssertEqual(ProviderKind.defaultForNewInstalls.defaultModel, "gemini-3.5-flash")
    }

    /// `displayName` names what ran, and it is written into history rows, log lines and connection
    /// errors. Advice belongs only on the row of a picker.
    func testAdviceStaysOutOfTheNameUsedForRecords() {
        for kind in ProviderKind.allCases {
            XCTAssertFalse(kind.displayName.contains("recommended"))
            XCTAssertTrue(kind.pickerLabel.hasPrefix(kind.displayName))
        }
        XCTAssertEqual(ProviderKind.google.pickerLabel, "Google — recommended")
        XCTAssertEqual(ProviderKind.xai.pickerLabel, "xAI — recommended")
        XCTAssertEqual(ProviderKind.openrouter.pickerLabel, "OpenRouter")
    }

    /// The iOS wording exists because iOS cannot read another app's screen. If it ever claimed the
    /// screen the note would be a promise the platform forbids — see `docs/PARITY.md`.
    func testTheUngroundedWordingNeverClaimsTheScreen() throws {
        let note = try XCTUnwrap(ProviderKind.google.ungroundedRecommendationNote)
        XCTAssertFalse(note.contains("reads the screen"))
        XCTAssertTrue(note.contains("audio alone"))
        // And the grounded one must, since that is the entire claim being made for it.
        let grounded = try XCTUnwrap(ProviderKind.google.recommendationNote)
        XCTAssertTrue(grounded.contains("reads the screen"))
        XCTAssertTrue(grounded.contains("seven recent jargon-heavy recordings"))
        XCTAssertTrue(grounded.contains("no human goldens"))
    }
}
